### PR TITLE
map2alm pixel weights

### PR DESCRIPTION
**Pull request by zonca**
Implementation of #356.

* Automatically download pixel weights file the first time `map2alm` is used with a specific nside using `astropy.data`
* Apply them before the call to the `map2alm` transform

### CODE REVIEW

**Comment by mreineck** [comment]
I'll check once I get home.

With pure C++, the largest deviations from 1 I see in the recovered a_lm are below 1e-9. So I suspect the problem lies in the interface or on the Python side.

**Comment by zonca** [comment]
also updated to the latest healpix trunk

**Comment by mreineck** [comment]
Sorry, I don't manage to get your branch compiled and set up properly... it keeps failing with
```
Sigma is 0.000000 arcmin (0.000000 rad) 
-> fwhm is 0.000000 arcmin
Traceback (most recent call last):
  File "healpy/test/test_pixelweights.py", line 30, in <module>
    a.test_pixelweights()
  File "healpy/test/test_pixelweights.py", line 23, in test_pixelweights
    alm = hp.map2alm(self.m, lmax=self.lmax, use_pixel_weights=True)
  File "/home/martin/codes/zhealpy/healpy/sphtfunc.py", line 167, in map2alm
    package="healpy")
  File "/home/martin/.local/lib/python2.7/site-packages/astropy/utils/data.py", line 636, in get_pkg_data_filename
    timeout=remote_timeout)
  File "/home/martin/.local/lib/python2.7/site-packages/astropy/utils/data.py", line 1074, in download_file
    if url_mirror in url2hash:
  File "/usr/lib/python2.7/shelve.py", line 110, in __contains__
    return key in self.dict
  File "/usr/lib/python2.7/_abcoll.py", line 388, in __contains__
    self[key]
  File "/usr/lib/python2.7/bsddb/__init__.py", line 270, in __getitem__
    return _DeadlockWrap(lambda: self.db[key])  # self.db[key]
  File "/usr/lib/python2.7/bsddb/dbutils.py", line 68, in DeadlockWrap
    return function(*_args, **_kwargs)
  File "/usr/lib/python2.7/bsddb/__init__.py", line 270, in <lambda>
    return _DeadlockWrap(lambda: self.db[key])  # self.db[key]
TypeError: String or Integer object expected for key, unicode found

```

But perhaps the  problem is simply the `apply_fullweights()` call ... this function actually changes its first argument. If a copy is made at the interface between both languages, you will most likely get your unchanged input back. Could you please check for this?

**Comment by zonca** [comment]
I see that the alm values change if I do the weighting or not, so I don't think that this is a problem.
I haven't tested the code in Python 2.7, I'll try now

**Comment by mreineck** [comment]
You need to set iter=0. Then it seems to work.

File: healpy/src/_sphtools.pyx [hunk 1]
```diff
@@ -301,6 +309,17 @@ def map2alm(m, lmax = None, mmax = None, niter = 3, use_weights = False,
     else:
         w_arr.allocAndFill(2 * nside, 1.)
 
+    cdef vector[double] full_weights
+    if pixel_weights_filename is not None:
+
+        full_weights = read_fullweights_from_fits(pixel_weights_filename.encode("UTF-8"), nside)
+        # pixel weighting requires 0 iterations
+        niter = 0
```
Comment (zonca) [inline healpy/src/_sphtools.pyx]: @mreineck do you mean `niter` here?

Comment (mreineck) [reply healpy/src/_sphtools.pyx]: Most likely yes. For my tests I simply called map2alm() in sphtfunc.py with iter=0.

**Comment by mreineck** [comment]
It's because a(l,-m) = (-1)**m conj(a(l,m)) (or something similar).
For m=0 this allows only real-valued a(l,m).

**Comment by zonca** [comment]
@lpsinger Python 2.7 fails on Travis-CI with the same Unicode error inside `astropy` @mreineck reported above, see https://travis-ci.org/healpy/healpy/jobs/382367728

The problem is that I tried with a conda Python 2.7 on my machine and it worked fine....any suggestion?

**Comment by mreineck** [comment]
This is actually a "feature" that worries me... do you really want a healpy functionality which only works when the machine is connected to the internet?

**Comment by zonca** [comment]
@mreineck there is no way around it, weights files are too big to be bundled inside the package.

**Comment by lpsinger** [comment]
> This is actually a "feature" that worries me... do you really want a healpy functionality which only works when the machine is connected to the internet?

Agreed. Why not distribute this with the package?

**Comment by mreineck** [comment]
I can provide them, if you like. I'll start the calculations tomorrow; no idea how long they will take however ...
The size of a weight file is rougly Npix/2 in (bytes).

**Comment by pllim** [comment]
What if... you add `from __future__ import unicode_literals` in your code and try it with Astropy 2.0.6? I think there is something about unicode vs str stuff going on.

**Comment by zonca** [comment]
thanks @pllim, it works now!

@lpsinger would you like to review this PR?

**Comment by zonca** [comment]
so 4096 should be ~100 MB, I think they are too big, I think using `astropy`'s utils.data to retrieve them is fine.

**Comment by zonca** [comment]
add `use_pixel_weights` also to `hp.anafast`

**Comment by pllim** [comment]
FWIW a fix was put in for the next astropy 2.0.7 release but please let us know if you still encounter this error for `astropy>=2.0.7`. I apologize for any inconvenience caused. xref astropy/astropy#7493

**Comment by zonca** [comment]
removed version pinning of `astropy`, now it should get `2.0.7` for Python 2.7

**Comment by pllim** [comment]
Great! All green now.
```
Searching for astropy==2.0.7
Best match: astropy 2.0.7
Adding astropy 2.0.7 to easy-install.pth file
```

**Pull request merged**